core/src/test/java/org/apache/iceberg/TestMetadataTableFilters.java

```diff
@@ -252,6 +254,74 @@ public void testPlanTasks() {
     validateCombinedScanTasks(tasks, 0);
   }
 
+  @Test
+  public void testPartitionSpecEvolution() {
+    Assume.assumeTrue(formatVersion == 2);
```

Comment:
It is pretty tricky to have the same code work in v1 and v2 to change spec and write new files, so choosing to do this in v2 (should be orthogonal to the problem).

Reply:
I'd probably prefer not to reuse `populateTableNewSpec` and duplicate some parts of the code but cover v1 and v2 tables to be safe. Using extra helper methods in tests reduces code duplication but makes it hard to review such tests as I had to jump to those methods to understand the current table state.

I'd also add tests that refer to the old partition spec. The `partition` column is a union of all partition types. We should push down conditions on `data_bucket` to manifests that belong to the old spec.